Website Attack Countermeasures

aszepieniec opened the pull request:
This PR
 - Adds a framework for simulating web-level attacks such as DoS through
   flooding and for monitoring the health of the server. The purpose of this
  framework is to quantify and test the efficacy of countermeasures against
  such attacks.
 - Adds a basic countermeasure against flooding, namely rate-limiting.

To test without rate-limiting:
 1. Open 4 terminal windows.
 2. Start neptune-core in one.
 3. Start neptune-explorer in two: `cargo run -- --site-domain test`
 4. Start the scraper in three: `cargo run --bin scraper --features="attacks"`
    and observe the green info logs indicating successful retrieval of web
    pages.
 5. Start the attack in four:
    `cargo run --bin high_rate_attack --features="attacks"`.
    Observe that in window three the green info messages turn into orange warn
    messages, indicating response timeouts. Also, try opening an explorer page.
    It gives a timeout.

To test with rate-limiting:
 3. Start neptune-explorer in two, but with feature flag "attacks" enabled:
    `cargo run --features="attacks" -- --site-domain test`. With this feature flag
    enabled, the value of HTTP header X-Real-IP-Override is used for the IP, so
    so you can simulate an attack from a different IP from the scraper.
 5. Start the attack and observe that the green info messages continue without
    interruption. Opening an explorer page does not result in a timeout.

aszepieniec committed:
security: Add dumb flooding attack

aszepieniec committed:
feat: Add rate-limiting as countermeasure against flooding

Add rate-limiting logic to give every IP a quotum of 1 request per 10
milliseconds. The IP is obtained from the socket address.

When the feature flag "attacks" is enabled on the server, and the HTTP header
field "X-Real-IP-Override" on the request is set, then the server uses the field
value instead of obtaining it from the socket address. This configuration allows
the rate-limiting logic to be tested.

To test without rate-limiting:
 1. Open 4 terminal windows.
 2. Start neptune-core in one.
 3. Start neptune-explorer in two: `cargo run -- --site-domain test`
 4. Start the scraper in three: `cargo run --bin scraper --features="attacks"`
    and observe the green info logs indicating successful retrieval of web
    pages.
 5. Start the attack in four:
    `cargo run --bin high_rate_attack --features="attacks"`.
    Observe that in window three the green info messages turn into orange warn
    messages, indicating response timeouts. Also, try opening an explorer page.
    It gives a timeout.

To test with rate-limiting:
 3. Start neptune-explorer in two, but with feature flag "attacks" enabled:
    `cargo run --features="attacks" -- --site-domain test`
 5. Start the attack and observe that the green info messages continue without
    interruption. Opening an explorer page does not result in a timeout.

jan-ferdinand reviewed (commented):
Some nits & questions.

aszepieniec commented on src/main.rs:
```diff
@@ -83,4 +95,61 @@ pub fn setup_routes(app_state: AppState) -> Router {
+        {
+            if let Some(ip_header_value) = headers.get("X-Real-IP-Override") {
+                if let Ok(ip_str) = ip_header_value.to_str() {
+                    if let Ok(ip) = IpAddr::from_str(ip_str) {
```
Well, the compiler did complain and did inform me not to worry because let-chains were going to be available in 2024. I'm less confused now; thanks!

From what I understand, relying on the earlier version of rust is a good thing because more compatible. So no need to upgrade, is that right? Regardless, even if upgrading to 2024 is deemed necessary or advisable, it is out of scope for this PR and I nominate the deemer for that task.

aszepieniec committed:
docs: Error-corrrect comment

The word "false", while spelled correctly, is grammatically incorrect. The word "fails" fits better semantically, is correct from a grammar point of view, and lives at a small edit distance (not to mention etymological distance) from "false".

Co-authored-by: Jan Ferdinand Sauer <ferdinand@neptune.cash>

jan-ferdinand commented on src/main.rs:
```diff
@@ -83,4 +95,61 @@ pub fn setup_routes(app_state: AppState) -> Router {
+            }
+        }
+
+        let ip = {
```
1. Agreed.
1. This stuff seems really complicated. I don’t claim to understand enough of internet architecture and HTTP headers to know the best approach here. Since you seem to have spend a bunch of thought on the matter, I trust your judgement.
